squirrel/static_analyser/analyser.cpp

```diff
@@ -1,62 +1,3581 @@
+
+  if (_equalChecker.check(ifTrue, ifFalse)) {
+    report(expr, DiagnosticsId::DI_TERNARY_SAME_VALUES);
+  }
+}
+
+void CheckerVisitor::checkAlwaysTrueOrFalse(const BinExpr *bin) {
+  enum TreeOp op = bin->op();
+  if (op != TO_ANDAND && op != TO_OROR)
+    return;
+
+  const Expr *lhs = maybeEval(bin->lhs())->asExpression();
+  const Expr *rhs = maybeEval(bin->rhs())->asExpression();
+
+  enum TreeOp cmpOp = lhs->op();
+  if (cmpOp == rhs->op() && (cmpOp == TO_NE || cmpOp == TO_EQ)) {
+    const char *constValue = nullptr;
+
+    if (op == TO_ANDAND && cmpOp == TO_EQ)
+      constValue = "false";
+
+    if (op == TO_OROR && cmpOp == TO_NE)
+      constValue = "true";
+
+    if (!constValue)
+      return;
+
+    const BinExpr *lhsBin = static_cast<const BinExpr *>(lhs);
+    const BinExpr *rhsBin = static_cast<const BinExpr *>(rhs);
+
+    const Expr *lconst = lhsBin->rhs();
+    const Expr *rconst = rhsBin->rhs();
+
+    if (lconst->op() == TO_LITERAL || isUpperCaseIdentifier(lconst)) {
+      if (rconst->op() == TO_LITERAL || isUpperCaseIdentifier(rconst)) {
+        if (_equalChecker.check(lhsBin->lhs(), rhsBin->lhs())) {
+          if (!_equalChecker.check(lconst, rconst)) {
+            report(bin, DiagnosticsId::DI_ALWAYS_T_OR_F, constValue);
+          }
+        }
+      }
+    }
+  }
+  else if (lhs->op() == TO_NOT || rhs->op() == TO_NOT && (lhs->op() != rhs->op())) {
+    const char *v = op == TO_OROR ? "true" : "false";
+    if (lhs->op() == TO_NOT) {
+      const UnExpr *u = static_cast<const UnExpr *>(lhs);
+      if (_equalChecker.check(u->argument(), rhs)) {
+        report(bin, DiagnosticsId::DI_ALWAYS_T_OR_F, v);
+      }
+    }
+    if (rhs->op() == TO_NOT) {
+      const UnExpr *u = static_cast<const UnExpr *>(rhs);
+      if (_equalChecker.check(lhs, u->argument())) {
+        report(bin, DiagnosticsId::DI_ALWAYS_T_OR_F, v);
+      }
+    }
+  }
+}
+
+void CheckerVisitor::checkDeclarationInArith(const BinExpr *bin) {
+  if (isArithOperator(bin->op())) {
+    const Expr *lhs = maybeEval(bin->lhs());
+    const Expr *rhs = maybeEval(bin->rhs());
+
+    if (lhs->op() == TO_DECL_EXPR || lhs->op() == TO_ARRAYEXPR) {
+      report(bin->lhs(), DiagnosticsId::DI_DECL_IN_EXPR);
+    }
+
+    if (rhs->op() == TO_DECL_EXPR || rhs->op() == TO_ARRAYEXPR) {
+      report(bin->rhs(), DiagnosticsId::DI_DECL_IN_EXPR);
+    }
+  }
+}
+
+void CheckerVisitor::checkIntDivByZero(const BinExpr *bin) {
+  if (isDivOperator(bin->op())) {
+    const Expr *divided = skipUnary(maybeEval(skipUnary(bin->lhs())));
+    const Expr *divisor = skipUnary(maybeEval(skipUnary(bin->rhs())));
+
+    if (divided->op() == TO_LITERAL && divisor->op() == TO_LITERAL) {
+      const LiteralExpr *l = divided->asLiteral();
+      const LiteralExpr *r = divisor->asLiteral();
+      if (l->kind() == LK_INT && r->kind() == LK_INT) {
+        if (r->raw() == 0) {
+          report(bin, DiagnosticsId::DI_DIV_BY_ZERO);
+        }
+        else if (l->i() % r->i() != 0) {
```

Comment:
also need check for (std::limits<typeof l->i()>::min() % -1) to avoid integer overflow and crash in analyzer

Reply:
done